File: dimos-ar/clients/web/src/protocol.ts
/** Keep in sync with dimos_ar/protocol.py and docs/PROTOCOL.md */

export const PROTOCOL_VERSION = 1;
/** Default for replay mode; live hardware uses hello.robots[0] after connect. */
export const ROBOT_ID = "go2";

let activeRobotId: string | null = null;

export function setActiveRobotId(robotId: string): void {
  activeRobotId = robotId;
}

export function clearActiveRobotId(): void {
  activeRobotId = null;
}

export function getActiveRobotId(): string | null {
  return activeRobotId;
}

function requireActiveRobotId(action: string): string {
  if (!activeRobotId) {
    throw new Error(`Cannot send ${action} before hello negotiates robot_id`);
  }
  return activeRobotId;
}

export interface HelloMessage {
  type: "hello";
  protocol_version: number;
  robots: string[];
  capabilities: string[];
}

export interface LidarMessage {
  type: "lidar";
  ts: number;
  robot_id: string;
  frame: string;
  points: [number, number, number][];
  colors?: [number, number, number][];
}

export interface PoseMessage {
  type: "pose";
  ts: number;
  robot_id: string;
  frame: string;
  position: [number, number, number];
  orientation: [number, number, number, number];
}

export interface RegisteredMessage {
  type: "registered";
  ts: number;
  robot_id: string;
  registered: boolean;
}

export interface AlignStatusMessage {
  type: "align_status";
  ts: number;
  robot_id: string;
  state: "detecting" | "aligned" | "failed";
  robot_marker_detected: boolean;
  spectacles_marker_detected: boolean;
  quality?: number;
  best_quality?: number;
  has_candidate?: boolean;
  candidate_count?: number;
  message: string;
}

export interface BridgeStatusMessage {
  type: "bridge_status";
  ts: number;
  robot_id: string;
  mode: "live" | "replay";
  robot_connected: boolean;
  robot_model: string;
  robot_serial?: string;
  streams_active: boolean;
  registered: boolean;
  reconnecting: boolean;
}

export type OutboundMessage =
  | HelloMessage
  | LidarMessage
  | PoseMessage
  | RegisteredMessage
  | AlignStatusMessage
  | BridgeStatusMessage;

export function buildGetStatus(): string {
  return JSON.stringify({
    type: "get_status",
    ts: Date.now() / 1000,
    robot_id: requireActiveRobotId("get_status"),
  });
}

export function buildDevRegister(): string {
  return JSON.stringify({
    type: "register",
    ts: Date.now() / 1000,
    robot_id: requireActiveRobotId("register"),
    marker_id: 0,
    marker_position: [0, 0, 0],
    marker_orientation: [0, 0, 0, 1],
  });
}

export function buildAlignCommit(): string {
  return JSON.stringify({
    type: "align_commit",
    ts: Date.now() / 1000,
    robot_id: requireActiveRobotId("align_commit"),
  });
}

function requireObject(data: unknown): Record<string, unknown> {
  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    throw new Error("Message must be a JSON object");
  }
  return data as Record<string, unknown>;
}

function requireString(obj: Record<string, unknown>, key: string): string {
  const v = obj[key];
  if (typeof v !== "string") {
    throw new Error(`Missing or invalid field: ${key}`);
  }
  return v;
}

function requireNumber(obj: Record<string, unknown>, key: string): number {
  const v = obj[key];
  if (typeof v !== "number") {
    throw new Error(`Missing or invalid field: ${key}`);
  }
  return v;
}

function parsePoints(raw: unknown): [number, number, number][] {
  if (!Array.isArray(raw)) {
    throw new Error("points must be an array");
  }
  return raw.map((p) => {
    if (!Array.isArray(p) || p.length !== 3) {
      throw new Error("each point must be [x, y, z]");
    }
    return [Number(p[0]), Number(p[1]), Number(p[2])] as [number, number, number];
  });
}

function parseVec3(raw: unknown): [number, number, number] {
  if (!Array.isArray(raw) || raw.length !== 3) {
    throw new Error("expected [x, y, z]");
  }
  return [Number(raw[0]), Number(raw[1]), Number(raw[2])];
}

function parseColors(raw: unknown): [number, number, number][] | undefined {
  if (raw === undefined) {
    return undefined;
  }
  if (!Array.isArray(raw)) {
    throw new Error("colors must be an array");
  }
  return raw.map((c) => {
    if (!Array.isArray(c) || c.length !== 3) {
      throw new Error("each color must be [r, g, b]");
    }
    return [Number(c[0]), Number(c[1]), Number(c[2])] as [number, number, number];
  });
}

function unflattenVec3(flat: number[]): [number, number, number][] {
  const out: [number, number, number][] = [];
  for (let i = 0; i + 2 < flat.length; i += 3) {
    out.push([flat[i], flat[i + 1], flat[i + 2]]);
  }
  return out;
}

export function parseOutboundMessage(text: string): OutboundMessage | null {
  const data = requireObject(JSON.parse(text));
  const type = requireString(data, "type");

  switch (type) {
    case "hello": {
      const msg: HelloMessage = {
        type: "hello",
        protocol_version: requireNumber(data, "protocol_version"),
        robots: (data.robots as unknown[]).map(String),
        capabilities: (data.capabilities as unknown[]).map(String),
      };
      if (msg.robots.length > 0) {
        setActiveRobotId(msg.robots[0]);
      }
      return msg;
    }
    case "lidar": {
      let pts: [number, number, number][];
      if (Array.isArray(data.points_flat)) {
        pts = unflattenVec3(data.points_flat as number[]);
      } else {
        pts = parsePoints(data.points);
      }
      const msg: LidarMessage = {
        type: "lidar",
        ts: requireNumber(data, "ts"),
        robot_id: requireString(data, "robot_id"),
        frame: requireString(data, "frame"),
        points: pts,
      };
      setActiveRobotId(msg.robot_id);
      if (Array.isArray(data.colors_flat)) {
        msg.colors = unflattenVec3(data.colors_flat as number[]);
      } else {
        const colors = parseColors(data.colors);
        if (colors !== undefined) {
          msg.colors = colors;
        }
      }
      return msg;
    }
    case "pose": {
      const msg: PoseMessage = {
        type: "pose",
        ts: requireNumber(data, "ts"),
        robot_id: requireString(data, "robot_id"),
        frame: requireString(data, "frame"),
        position: parseVec3(data.position),
        orientation: (() => {
          const q = data.orientation;
          if (!Array.isArray(q) || q.length !== 4) {
            throw new Error("orientation must be [qx, qy, qz, qw]");
          }
          return [Number(q[0]), Number(q[1]), Number(q[2]), Number(q[3])] as [
            number,
            number,
            number,
            number,
          ];
        })(),
      };
      setActiveRobotId(msg.robot_id);
      return msg;
    }
    case "registered": {
      const msg: RegisteredMessage = {
        type: "registered",
        ts: requireNumber(data, "ts"),
        robot_id: requireString(data, "robot_id"),
        registered: Boolean(data.registered),
      };
      setActiveRobotId(msg.robot_id);
      return msg;
    }
    case "align_status": {
      const state = requireString(data, "state");
      if (state !== "detecting" && state !== "aligned" && state !== "failed") {
        throw new Error("align_status.state must be detecting, aligned, or failed");
      }
      const msg: AlignStatusMessage = {
        type: "align_status",
        ts: requireNumber(data, "ts"),
        robot_id: requireString(data, "robot_id"),
        state,
        robot_marker_detected: Boolean(data.robot_marker_detected),
        spectacles_marker_detected: Boolean(data.spectacles_marker_detected),
        message: typeof data.message === "string" ? data.message : "",
      };
      if (typeof data.quality === "number") {
        msg.quality = data.quality;
      }
      if (typeof data.best_quality === "number") {
        msg.best_quality = data.best_quality;
      }
      if (typeof data.has_candidate === "boolean") {
        msg.has_candidate = data.has_candidate;
      }
      if (typeof data.candidate_count === "number") {
        msg.candidate_count = data.candidate_count;
      }
      setActiveRobotId(msg.robot_id);
      return msg;
    }
    case "bridge_status": {
      const mode = requireString(data, "mode");
      if (mode !== "live" && mode !== "replay") {
        throw new Error("bridge_status.mode must be live or replay");
      }
      const msg: BridgeStatusMessage = {
        type: "bridge_status",
        ts: requireNumber(data, "ts"),
        robot_id: requireString(data, "robot_id"),
        mode,
        robot_connected: Boolean(data.robot_connected),
        robot_model: requireString(data, "robot_model"),
        streams_active: Boolean(data.streams_active),
        registered: Boolean(data.registered),
        reconnecting: Boolean(data.reconnecting),
      };
      if (typeof data.robot_serial === "string") {
        msg.robot_serial = data.robot_serial;
      }
      setActiveRobotId(msg.robot_id);
      return msg;
    }
    default:
      return null;
  }
}

export function formatBridgeStatus(msg: BridgeStatusMessage): string {
  const model = msg.robot_model.replace("unitree_", "").toUpperCase();
  if (msg.reconnecting) {
    return `Reconnecting to ${model}…`;
  }
  if (msg.mode === "replay") {
    return `Replay (${model})`;
  }
  const serial = msg.robot_serial ?? msg.robot_id;
  const streams = msg.streams_active ? "streams OK" : "waiting for streams";
  const reg = msg.registered ? "registered" : "not registered";
  return `${model} ${serial.slice(0, 8)}… · ${streams} · ${reg}`;
}

File: dimos-ar/dimos_ar/protocol.py
"""WebSocket JSON protocol — keep in sync with docs/PROTOCOL.md."""

from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import Any

import numpy as np
from numpy.typing import NDArray

from dimos_ar.bridge_status import BridgeStatusSnapshot

PROTOCOL_VERSION = 1
ROBOT_ID = "go2"
FRAME_WORLD = "world"

DEFAULT_CAPABILITIES = ["lidar", "odom", "align", "align_manual", "nav", "path", "emergency_stop"]


def _dumps(payload: dict[str, Any]) -> str:
    return json.dumps(payload, separators=(",", ":"), allow_nan=False)


@dataclass(frozen=True)
class RegisterMessage:
    ts: float
    robot_id: str
    marker_id: int
    marker_position: tuple[float, float, float]
    marker_orientation: tuple[float, float, float, float]


@dataclass(frozen=True)
class NavGoalMessage:
    ts: float
    robot_id: str
    position: tuple[float, float, float]
    orientation: tuple[float, float, float, float] | None = None


@dataclass(frozen=True)
class CancelGoalMessage:
    ts: float
    robot_id: str


@dataclass(frozen=True)
class EmergencyStopMessage:
    ts: float
    robot_id: str


@dataclass(frozen=True)
class AlignStartMessage:
    ts: float
    robot_id: str


@dataclass(frozen=True)
class AlignStopMessage:
    ts: float
    robot_id: str


@dataclass(frozen=True)
class AlignCommitMessage:
    ts: float
    robot_id: str


@dataclass(frozen=True)
class AlignMarkerMessage:
    ts: float
    robot_id: str
    marker_position: tuple[float, float, float]
    marker_orientation: tuple[float, float, float, float]


@dataclass(frozen=True)
class AlignManualPoseMessage:
    ts: float
    robot_id: str
    position: tuple[float, float, float]
    orientation: tuple[float, float, float, float]


@dataclass(frozen=True)
class GetStatusMessage:
    ts: float
    robot_id: str


InboundMessage = (
    RegisterMessage
    | NavGoalMessage
    | CancelGoalMessage
    | EmergencyStopMessage
    | AlignStartMessage
    | AlignStopMessage
    | AlignCommitMessage
    | AlignMarkerMessage
    | AlignManualPoseMessage
    | GetStatusMessage
)


def _require_type(data: dict[str, Any], key: str, expected: type) -> Any:
    if key not in data:
        raise ValueError(f"Missing required field: {key}")
    value = data[key]
    if not isinstance(value, expected):
        raise ValueError(f"Field {key!r} must be {expected.__name__}, got {type(value).__name__}")
    return value


def _vec3(data: dict[str, Any], key: str) -> tuple[float, float, float]:
    raw = _require_type(data, key, list)
    if len(raw) != 3:
        raise ValueError(f"Field {key!r} must be a 3-element array")
    return float(raw[0]), float(raw[1]), float(raw[2])


def _quat(data: dict[str, Any], key: str) -> tuple[float, float, float, float]:
    raw = _require_type(data, key, list)
    if len(raw) != 4:
        raise ValueError(f"Field {key!r} must be a 4-element quaternion [qx, qy, qz, qw]")
    return float(raw[0]), float(raw[1]), float(raw[2]), float(raw[3])


def _parse_marker_id(data: dict[str, Any]) -> int:
    if "marker_id" not in data:
        raise ValueError("Missing required field: marker_id")
    raw = data["marker_id"]
    if isinstance(raw, bool):
        raise ValueError("Field 'marker_id' must be an integer, got bool")
    if isinstance(raw, int):
        return raw
    if isinstance(raw, float):
        if raw.is_integer():
            return int(raw)
        raise ValueError("Field 'marker_id' must be a whole number")
    raise ValueError(
        f"Field 'marker_id' must be int, got {type(raw).__name__}",
    )


def decode_inbound(text: str, *, expected_robot_id: str | None = None) -> InboundMessage:
    """Parse an inbound JSON message. Raises ValueError on malformed input."""
    data = json.loads(text)
    if not isinstance(data, dict):
        raise ValueError("Message must be a JSON object")
    msg_type = _require_type(data, "type", str)
    if "ts" not in data or not isinstance(data["ts"], (int, float)):
        raise ValueError("Missing or invalid field: ts")
    ts = float(data["ts"])
    robot_id = str(_require_type(data, "robot_id", str))
    if expected_robot_id is not None and robot_id != expected_robot_id:
        raise ValueError(
            f"Unknown robot_id {robot_id!r}, expected {expected_robot_id!r}",
        )

    if msg_type == "register":
        return RegisterMessage(
            ts=ts,
            robot_id=robot_id,
            marker_id=_parse_marker_id(data),
            marker_position=_vec3(data, "marker_position"),
            marker_orientation=_quat(data, "marker_orientation"),
        )
    if msg_type == "nav_goal":
        orientation = _quat(data, "orientation") if "orientation" in data else None
        return NavGoalMessage(
            ts=ts,
            robot_id=robot_id,
            position=_vec3(data, "position"),
            orientation=orientation,
        )
    if msg_type == "cancel_goal":
        return CancelGoalMessage(ts=ts, robot_id=robot_id)
    if msg_type == "emergency_stop":
        return EmergencyStopMessage(ts=ts, robot_id=robot_id)
    if msg_type == "align_start":
        return AlignStartMessage(ts=ts, robot_id=robot_id)
    if msg_type == "align_stop":
        return AlignStopMessage(ts=ts, robot_id=robot_id)
    if msg_type == "align_commit":
        return AlignCommitMessage(ts=ts, robot_id=robot_id)
    if msg_type == "align_marker":
        return AlignMarkerMessage(
            ts=ts,
            robot_id=robot_id,
            marker_position=_vec3(data, "marker_position"),
            marker_orientation=_quat(data, "marker_orientation"),
        )
    if msg_type == "align_manual_pose":
        return AlignManualPoseMessage(
            ts=ts,
            robot_id=robot_id,
            position=_vec3(data, "position"),
            orientation=_quat(data, "orientation"),
        )
    if msg_type == "get_status":
        return GetStatusMessage(ts=ts, robot_id=robot_id)
    raise ValueError(f"Unknown inbound message type: {msg_type!r}")


def encode_bridge_status(
    snapshot: BridgeStatusSnapshot,
    *,
    ts: float | None = None,
) -> str:
    payload: dict[str, Any] = {
        "type": "bridge_status",
        "ts": ts if ts is not None else time.time(),
        "robot_id": snapshot.robot_id,
        "mode": snapshot.mode,
        "robot_connected": snapshot.robot_connected,
        "robot_model": snapshot.robot_model,
        "streams_active": snapshot.streams_active,
        "registered": snapshot.registered,
        "reconnecting": snapshot.reconnecting,
    }
    if snapshot.robot_serial is not None:
        payload["robot_serial"] = snapshot.robot_serial
    if snapshot.registration_method is not None:
        payload["registration_approximate"] = snapshot.registration_approximate
    return _dumps(payload)


def encode_hello(
    capabilities: list[str] | None = None,
    *,
    robot_id: str = ROBOT_ID,
) -> str:
    caps = capabilities if capabilities is not None else DEFAULT_CAPABILITIES
    return _dumps(
        {
            "type": "hello",
            "protocol_version": PROTOCOL_VERSION,
            "robots": [robot_id],
            "capabilities": caps,
        }
    )


def encode_registered(
    *,
    registered: bool = True,
    ts: float | None = None,
    robot_id: str = ROBOT_ID,
) -> str:
    return _dumps(
        {
            "type": "registered",
            "ts": ts if ts is not None else time.time(),
            "robot_id": robot_id,
            "registered": registered,
        }
    )


def _round_flat(arr: NDArray[np.floating], decimals: int = 3) -> list[float]:
    """Flatten Nx3 array to a 1-D list of rounded floats for compact JSON."""
    arr_f64 = arr.astype(np.float64)
    # Replace NaN/Inf with 0 to ensure valid JSON
    arr_f64 = np.nan_to_num(arr_f64, nan=0.0, posinf=0.0, neginf=0.0)
    return np.round(arr_f64, decimals).ravel().tolist()


def encode_lidar(
    *,
    ts: float,
    points: NDArray[np.floating],
    colors: NDArray[np.floating] | None = None,
    robot_id: str = ROBOT_ID,
) -> str:
    payload: dict[str, Any] = {
        "type": "lidar",
        "ts": round(ts, 3),
        "robot_id": robot_id,
        "frame": FRAME_WORLD,
        "points_flat": _round_flat(points),
    }
    if colors is not None and len(colors) > 0:
        payload["colors_flat"] = _round_flat(colors)
    return _dumps(payload)


def encode_align_status(
    *,
    ts: float | None = None,
    robot_id: str = ROBOT_ID,
    state: str = "detecting",
    robot_marker_detected: bool = False,
    spectacles_marker_detected: bool = False,
    quality: float | None = None,
    best_quality: float | None = None,
    has_candidate: bool | None = None,
    method: str | None = None,
    message: str = "",
) -> str:
    payload: dict[str, Any] = {
        "type": "align_status",
        "ts": ts if ts is not None else time.time(),
        "robot_id": robot_id,
        "state": state,
        "robot_marker_detected": robot_marker_detected,
        "spectacles_marker_detected": spectacles_marker_detected,
        "message": message,
    }
    if quality is not None:
        payload["quality"] = quality
    if best_quality is not None:
        payload["best_quality"] = best_quality
    if has_candidate is not None:
        payload["has_candidate"] = has_candidate
    if method is not None:
        payload["method"] = method
    return _dumps(payload)


def encode_pose(
    *,
    ts: float,
    position: tuple[float, float, float],
    orientation: tuple[float, float, float, float],
    robot_id: str = ROBOT_ID,
) -> str:
    return _dumps(
        {
            "type": "pose",
            "ts": ts,
            "robot_id": robot_id,
            "frame": FRAME_WORLD,
            "position": list(position),
            "orientation": list(orientation),
        }
    )


def encode_path(
    *,
    ts: float,
    waypoints: list[tuple[float, float, float]],
    robot_id: str = ROBOT_ID,
) -> str:
    return _dumps(
        {
            "type": "path",
            "ts": ts,
            "robot_id": robot_id,
            "frame": FRAME_WORLD,
            "waypoints": [list(point) for point in waypoints],
        }
    )


def encode_nav_status(
    *,
    ts: float | None = None,
    state: str,
    goal_reached: bool,
    robot_id: str = ROBOT_ID,
) -> str:
    return _dumps(
        {
            "type": "nav_status",
            "ts": ts if ts is not None else time.time(),
            "robot_id": robot_id,
            "state": state,
            "goal_reached": goal_reached,
        }
    )
